src/coreclr/jit/morph.cpp

```diff
@@ -323,6 +323,222 @@ GenTree* Compiler::fgMorphExpandCast(GenTreeCast* tree)
             }
         }
     }
+
+    // This if check needs to be changed to make sure we only
+    // block casts which are already Fixed UP.
+    do
```

Comment:
Why does this need to be `do..while(false)`?

Reply:
I was to break out of the checks in case where ```tree->IsSaturatedConversion()``` is set. So this was mainly to handle recursion. Since we are trying to get rid of the recursion in all, this would be removed if we remove the recursion.

Reply:
Right. I think this all stems from the recursive `fgMorphTree` call, which should be unnecessary since we're doing the morph and introducing the already correct nodes.

Reply:
This would notably all be resolved if this was done in lowering instead so that all we saw in the high level IR was cast and the backend is what made it correct for x86/x64, but that's more involved change.